### PR TITLE
Storing byte code size in the byte code header.

### PR DESCRIPTION
**Pull request by zherczeg**
This reduces the memory consumption, because the new allocator uses less memory if
the size as available when a block is freed. Snapshot generation is also simplified.

### CODE REVIEW

File: jerry-core/ecma/base/ecma-helpers.c [hunk 1]
```diff
@@ -1316,12 +1316,12 @@ void
 ecma_bytecode_ref (ecma_compiled_code_t *bytecode_p) /**< byte code pointer */
 {
   /* Abort program if maximum reference number is reached. */
-  if ((bytecode_p->status_flags >> ECMA_BYTECODE_REF_SHIFT) >= 0x3ff)
+  if (bytecode_p->refs >= 0xffff)
```
Comment (franc0is) [inline jerry-core/ecma/base/ecma-helpers.c]: use `UINT16_MAX`

File: jerry-core/ecma/base/ecma-globals.h [hunk 2]
```diff
@@ -741,18 +741,12 @@ typedef uintptr_t ecma_external_pointer_t;
   */
 typedef struct
 {
-  uint16_t status_flags;            /**< various status flags */
+  uint16_t size;                    /**< real size >> MEM_ALIGNMENT_LOG */
+  uint16_t refs;                    /**< reference counter for the byte code */
+  uint16_t status_flags;            /**< various status flags. */
```
Comment (franc0is) [inline jerry-core/ecma/base/ecma-globals.h]: Are the flags used for anything after you broke out the references? If not, why keep the field at all?

File: jerry-core/jerry.c [hunk 3]
```diff
@@ -1845,81 +1828,112 @@ snapshot_add_compiled_code (ecma_compiled_code_t *compiled_code_p, /**< compiled
...
+    return 0;
   }
 
-  snapshot_last_compiled_code_offset = snapshot_buffer_write_offset;
+  uint16_t start_offset = (uint16_t) (snapshot_buffer_write_offset >> MEM_ALIGNMENT_LOG);
+  ecma_compiled_code_t *copied_compiled_code_p;
 
-  compiled_code_map_entry_t *new_entry;
-  new_entry = (compiled_code_map_entry_t *) mem_heap_alloc_block (sizeof (compiled_code_map_entry_t));
+  copied_compiled_code_p = (ecma_compiled_code_t *) (snapshot_buffer_p + snapshot_buffer_write_offset);
```
Comment (franc0is) [inline jerry-core/jerry.c]: declare and define on a single line

File: jerry-core/jerry.c [hunk 4]
```diff
@@ -1845,81 +1828,112 @@ snapshot_add_compiled_code (ecma_compiled_code_t *compiled_code_p, /**< compiled
...
-      return;
     }
 
-    data_p = regexp_pattern;
+    MEM_FINALIZE_LOCAL_ARRAY (buffer_p);
+
+    snapshot_buffer_write_offset = JERRY_ALIGNUP (snapshot_buffer_write_offset, MEM_ALIGNMENT);
+
+    /* Regexp character size is stored in refs. */
+    copied_compiled_code_p->refs = (uint16_t) pattern_size;
```
Comment (franc0is) [inline jerry-core/jerry.c]: This seems to rely on the re_compiled_code_t and ecma_compiled_code_t structs to stay stable relative to each other. Why not create a new re_compiled_code_t variable pointing at the same memory instead?

File: jerry-core/jerry.c [hunk 5]
```diff
@@ -2079,64 +2047,56 @@ jerry_parse_and_save_snapshot (const jerry_api_char_t *source_p, /**< script sou
...
+    return 0;
+  }
 
-      jrt_write_to_buffer_by_offset (buffer_p,
-                                     buffer_size,
-                                     &header_offset,
-                                     &header,
-                                     sizeof (header));
+  jerry_snapshot_header_t header;
+  header.version = JERRY_SNAPSHOT_VERSION;
```
Comment (franc0is) [inline jerry-core/jerry.c]: Any reason why you wouldn't prefer struct initilization here?
```
jerry_snapshot_header_t header = {
  .version = JERRY_SNAPSHOT_VERSION,
  ...
};
```

It has the advantage that new fields added to the structure are zeroed out rather than left uninitialized.

File: jerry-core/jerry.c [hunk 6]
```diff
@@ -2213,21 +2169,25 @@ snapshot_load_compiled_code (const uint8_t *snapshot_data_p, /**< snapshot data
     header_size = sizeof (cbc_uint8_arguments_t);
   }
 
-  if (copy_bytecode)
+  if (copy_bytecode
+      || (header_size + (literal_end * sizeof (uint16_t)) + 8 > code_size))
```
Comment (franc0is) [inline jerry-core/jerry.c]: can something more descriptive be used instead of `+ 8` ?

File: jerry-core/parser/js/byte-code.h [hunk 7]
```diff
@@ -617,6 +617,8 @@
  */
 typedef struct
 {
+  uint16_t size;                    /**< real size >> MEM_ALIGNMENT_LOG */
```
Comment (franc0is) [inline jerry-core/parser/js/byte-code.h]: why not embed an `ecma_compiled_code_t` in these structs instead of copying the structure?

File: jerry-core/parser/js/js-parser-limits.h [hunk 8]
```diff
@@ -54,7 +54,7 @@
 /* Maximum code size.
  * Limit: 16777215. Recommended: 65535, 16777215. */
 #ifndef PARSER_MAXIMUM_CODE_SIZE
-#define PARSER_MAXIMUM_CODE_SIZE 16777215
+#define PARSER_MAXIMUM_CODE_SIZE (65535 << (MEM_ALIGNMENT_LOG))
```
Comment (franc0is) [inline jerry-core/parser/js/js-parser-limits.h]: Looks suspiciously like UINT16_MAX

File: jerry-core/ecma/base/ecma-globals.h (hunk 2 above)
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-globals.h]: The flags contains whether the byte code is function or regexp, so it is still important. The other flag bits depends on this.

File: jerry-core/jerry.c (hunk 3 above)
Comment (zherczeg) [inline jerry-core/jerry.c]: Would break the 120 character maximum line length. Other kind of line breaking is also not allowed so this is our only option currently.

File: jerry-core/jerry.c (hunk 4 above)
Comment (zherczeg) [inline jerry-core/jerry.c]: I don't understand this.

File: jerry-core/jerry.c (hunk 5 above)
Comment (zherczeg) [inline jerry-core/jerry.c]: Unfortunately you have to put { to the next line and I don't like that form. There cannot be any fields which are not initialized of this structure.

File: jerry-core/parser/js/byte-code.h (hunk 7 above)
Comment (zherczeg) [inline jerry-core/parser/js/byte-code.h]: This could be done in a follow-up patch. This patch is big enough, and I am sure such a change will require several line changes.

File: jerry-core/parser/js/js-parser-limits.h (hunk 8 above)
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-limits.h]: In this file numbers are explicit, so I prefer it this way.

**Comment by LaszloLango** [comment]
@zherczeg, please rebase to the current master

**Comment by zherczeg** [comment]
Raspberry Pi2 results:

                      Benchmark |          RSS(+ is better) |                   Perf(+ is better)
                     3d-cube.js |   80k ->   76k :  +5.000% |   2.530s ->   2.552s :  -0.870%  (+-0.892%) : [≈]
         access-binary-trees.js |   80k ->   76k :  +5.000% |   1.334s ->   1.348s :  -1.049%  (+-1.098%) : [≈]
             access-fannkuch.js |   40k ->   32k : +20.000% |   7.546s ->   7.596s :  -0.663%  (+-1.132%) : [≈]
                access-nbody.js |   40k ->   36k : +10.000% |   2.914s ->   2.920s :  -0.206%  (+-0.806%) : [≈]
    bitops-3bit-bits-in-byte.js |   28k ->   24k : +14.286% |   1.666s ->   1.684s :  -1.080%  (+-0.963%) : [-]
         bitops-bits-in-byte.js |   28k ->   24k : +14.286% |   2.420s ->   2.472s :  -2.149%  (+-0.932%) : [-]
          bitops-bitwise-and.js |   32k ->   24k : +25.000% |   3.290s ->   3.260s :  +0.912%  (+-1.318%) : [≈]
          bitops-nsieve-bits.js |  152k ->  148k :  +2.632% |  20.138s ->  20.132s :  +0.030%  (+-1.035%) : [≈]
       controlflow-recursive.js |  140k ->  136k :  +2.857% |   0.926s ->   0.932s :  -0.648%  (+-1.578%) : [≈]
                  crypto-aes.js |  108k ->  104k :  +3.704% |   3.544s ->   3.502s :  +1.185%  (+-0.575%) : [+]
                  crypto-md5.js |  176k ->  164k :  +6.818% |   7.128s ->   7.180s :  -0.730%  (+-1.304%) : [≈]
                 crypto-sha1.js |  124k ->  116k :  +6.452% |   3.884s ->   3.890s :  -0.154%  (+-0.712%) : [≈]
           date-format-tofte.js |   64k ->   60k :  +6.250% |   2.152s ->   2.168s :  -0.743%  (+-1.488%) : [≈]
           date-format-xparb.js |   56k ->   52k :  +7.143% |   0.856s ->   0.848s :  +0.935%  (+-2.399%) : [≈]
                 math-cordic.js |   36k ->   32k : +11.111% |   2.898s ->   2.880s :  +0.621%  (+-1.360%) : [≈]
           math-partial-sums.js |   32k ->   24k : +25.000% |   1.584s ->   1.574s :  +0.631%  (+-1.004%) : [≈]
          math-spectral-norm.js |   36k ->   32k : +11.111% |   1.430s ->   1.432s :  -0.140%  (+-0.644%) : [≈]
               string-base64.js |  156k ->  152k :  +2.564% |  30.992s ->  30.930s :  +0.200%  (+-0.876%) : [≈]
                string-fasta.js |   40k ->   36k : +10.000% |   3.504s ->   3.386s :  +3.368%  (+-1.091%) : [+]
                Geometric mean: |    RSS reduction: 10.232% |                Speed up: -0.022% (+-0.273%) : [≈]

## Binary sizes ##
master: 183,388
branch: 182,948

File: jerry-core/parser/js/byte-code.h (hunk 7 above)
Comment (zherczeg) [inline jerry-core/parser/js/byte-code.h]: Anyway, this is a good idea.

**Comment by LaszloLango** [comment]
I'd use the `'u'`suffix for number constants for unsigned variables and return types. LGTM besides that. Nice improvement.

File: jerry-core/jerry-snapshot.h [hunk 9]
```diff
@@ -24,27 +24,17 @@
...
 typedef struct
 {
-  uint32_t last_compiled_code_offset; /**< offset of the last compiled code */
+  uint32_t version; /**< version number */
+  uint32_t lit_table_offset; /**< offset of the literal table */
   uint32_t lit_table_size; /**< size of literal table */
-  __extension__ uint32_t is_run_global : 1; /**< flag, indicating whether the snapshot
-                                             *   was dumped as 'Global scope'-mode code (true)
-                                             *   or as eval-mode code (false) */
+  uint8_t is_run_global; /**< flag, indicating whether the snapshot
```
Comment (akiss77) [inline jerry-core/jerry-snapshot.h]: Why not simply use the type `bool` here? I don't see any alignment requirements or need for multiple flags that would necessitate an integer type with fixed bit-length.

File: jerry-core/jerry.c [hunk 10]
```diff
@@ -2023,42 +2019,19 @@ jerry_parse_and_save_snapshot (const jerry_api_char_t *source_p, /**< script sou
...
-  snapshot_report_byte_code_compilation = true;
   jerry_api_object_t *error_obj_p = NULL;
 
   if (is_for_global)
   {
-    parse_status = parser_parse_script (source_p, source_size, &bytecode_data_p, &error_obj_p);
+    parse_status = parser_parse_script (source_p,
+                                        source_size,
+                                        &bytecode_data_p,
+                                        &error_obj_p);
```
Comment (akiss77) [inline jerry-core/jerry.c]: Did anything else change here other than whitespace? If no, why the change then? The original code still seemed to fit within the 120 chars-per-line limit.

File: jerry-core/jerry-snapshot.h (hunk 9 above)
Comment (zherczeg) [inline jerry-core/jerry-snapshot.h]: I don't like using bools in structures because its storage size is unclear to me.

File: jerry-core/jerry.c (hunk 10 above)
Comment (zherczeg) [inline jerry-core/jerry.c]: Because parser_parse_eval below uses this style, and for me this looked ugly.

File: jerry-core/jrt/jrt.h [hunk 11]
```diff
@@ -204,7 +204,7 @@ extern void __noreturn jerry_fatal (jerry_fatal_code_t);
  *
  * Returns minimum positive value, that divides @a alignment and is more than or equal to @a value
  */
-#define JERRY_ALIGNUP(value, alignment) ((alignment) * (((value) + (alignment) - 1) / (alignment)))
+#define JERRY_ALIGNUP(value, alignment) (((value) + ((alignment) - 1)) & ~((alignment) - 1))
```
Comment (akiss77) [inline jerry-core/jrt/jrt.h]: This works only if `alignment` is a power of two, while original version worked for any positive integer number. True enough, power-of-2 alignments are used in practice, but then please update the doc accordingly.

File: jerry-core/parser/js/common.h [hunk 12]
```diff
@@ -52,11 +52,11 @@
...
 
-#define PARSER_MALLOC(size) mem_heap_alloc_block_store_size (size)
-#define PARSER_FREE(ptr) mem_heap_free_block_size_stored ((void *) ptr)
+#define PARSER_MALLOC(size) mem_heap_alloc_block (size)
+#define PARSER_FREE(ptr, size) mem_heap_free_block ((ptr), (size))
 
-#define PARSER_MALLOC_LOCAL(size) mem_heap_alloc_block_store_size (size)
-#define PARSER_FREE_LOCAL(ptr) mem_heap_free_block_size_stored (ptr)
+#define PARSER_MALLOC_LOCAL(size) mem_heap_alloc_block (size)
+#define PARSER_FREE_LOCAL(ptr, size) mem_heap_free_block ((ptr), (size))
```
Comment (akiss77) [inline jerry-core/parser/js/common.h]: Do we still need to keep these macros?
- They seem to duplicate each other.
- Their only valid uses seem to be in js-parser-mem.c.
- This patch already replaces them at some points.

I'd vote for replacing them with calls to `parser__{malloc,free}[_local]` or `mem_heap_{alloc,free}_block`, depending on the use case.

File: jerry-core/jerry.c (hunk 10 above)
Comment (akiss77) [inline jerry-core/jerry.c]: _This_ style takes too much vertical space from my screen for a single statement but _de gustibus non est disputandum_.

File: jerry-core/jerry-snapshot.h (hunk 9 above)
Comment (akiss77) [inline jerry-core/jerry-snapshot.h]: First I thought that exact size does not matter here since the struct is not used that often. Now I realize that it does matter, since this header becomes part of the snapshot buffer and will be written to disk, so we will need to know the exact offsets of fields and data, etc. That's really a point against `bool`.

However, there might still be some issues with `uint_8` and also with the way the snapshot buffer is constructed (in jerry.c). The size of a struct may differ from machine to machine, AFAIK it's up to the compiler to define alignments, padding, whatnot. Would this struct be packed, then its size would be 13 bytes, on a regular 32 bit machine it would most probably be 16 bytes in unpacked form, but actually there is no guarantee for that (and there may be exotic architectures where we might get a different result). Making the last field `uint32_t` makes things a bit more portable IMHO.

As for my just-realized concerns with jerry.c, I'll make a comment there.

File: jerry-core/jerry.c [hunk 13]
```diff
@@ -2023,42 +2019,19 @@ jerry_parse_and_save_snapshot (const jerry_api_char_t *source_p, /**< script sou
   jsp_status_t parse_status;
   ecma_compiled_code_t *bytecode_data_p;
 
-  snapshot_buffer_write_offset = 0;
-  snapshot_last_compiled_code_offset = 0;
+  snapshot_buffer_write_offset = JERRY_ALIGNUP (sizeof (jerry_snapshot_header_t),
+                                                MEM_ALIGNMENT);
```
Comment (akiss77) [inline jerry-core/jerry.c]: As we construct the snapshot buffer, we seem to use alignments based on `MEM_ALIGNMENT` and `MEM_ALIGNMENT_LOG` (scattered all around the code, not just here). Do we realize that changing those defines invalidates all our snapshots? I mean, there is a `JERRY_SNAPSHOT_VERSION` macro which is incremented every time there is a breaking change to the snapshot format (just like this patch). Memory alignment seems only to be a memory configuration parameter, I'm not sure it should have any effect on the snapshot format.

File: jerry-core/jerry.c [hunk 14]
```diff
@@ -1841,81 +1824,117 @@ snapshot_add_compiled_code (ecma_compiled_code_t *compiled_code_p, /**< compiled
...
-  snapshot_last_compiled_code_offset = snapshot_buffer_write_offset;
+  uint16_t start_offset = (uint16_t) (snapshot_buffer_write_offset >> MEM_ALIGNMENT_LOG);
+  ecma_compiled_code_t *copied_compiled_code_p;
 
-  compiled_code_map_entry_t *new_entry;
-  new_entry = (compiled_code_map_entry_t *) mem_heap_alloc_block (sizeof (compiled_code_map_entry_t));
+  copied_compiled_code_p = (ecma_compiled_code_t *) (snapshot_buffer_p + snapshot_buffer_write_offset);
 
-  if (new_entry == NULL)
+  if (!(compiled_code_p->status_flags & CBC_CODE_FLAGS_FUNCTION))
```
Comment (akiss77) [inline jerry-core/jerry.c]: The issue that @franc0is also mentioned is that within the body of this `if` we know that we are actually not dealing with `ecma_compiled_code_t` but `re_compiled_code_t`. It would be much comprehendible (and even safer) to access whatever fields we access as fields of `re_compiled_code_t` (by casting `compiled_code_p`).

Moreover, it should be noted at the struct declarations that the first 3 fields _must_ always be kept in sync.

File: jerry-core/ecma/base/ecma-globals.h (hunk 2 above)
Comment (akiss77) [inline jerry-core/ecma/base/ecma-globals.h]: The doc of this field should be amended with some short description of the valid flags (e.g., "various status flags (CBC_CODE_FLAGS_xxx)")

File: jerry-core/jerry.c (hunk 13 above)
Comment (zherczeg) [inline jerry-core/jerry.c]: This simplifies things. There are much more complex things we should worry about.

File: jerry-core/jerry.c (hunk 14 above)
Comment (zherczeg) [inline jerry-core/jerry.c]: No, saving a regex is not saving its byte code. We are only saving the ecma_compiled_code_t as a header and the pattern string. So we are not dealing with re_compiled_code_t.

**Comment by zherczeg** [comment]
Issues fixed.

File: jerry-core/jerry.c (hunk 13 above)
Comment (zherczeg) [inline jerry-core/jerry.c]: Basically all offsets in the snapshot are shifted with MEM_ALIGNMENT_LOG. We could support different alignments and other endian supports, but I would prefer that in a separate patch. I think we should evaluate the overhead first (whether it is worth it).

File: jerry-core/jerry.c (hunk 14 above)
Comment (akiss77) [inline jerry-core/jerry.c]: At line 1847 we have `((re_compiled_code_t *) compiled_code_p)->pattern_cp`, which hints that `compiled_code_p` is a regex compiled code. However, it's true that `copied_compiled_code_p` is an `ecma_compiled_code_t`. Since it seems that `compiled_code_p` is only used once, leave it as is.

File: jerry-core/parser/js/common.c [hunk 15]
```diff
@@ -165,7 +165,7 @@ util_free_literal (lexer_literal_t *literal_p) /**< literal */
   {
     if (!(literal_p->status_flags & LEXER_FLAG_SOURCE_PTR))
     {
-      PARSER_FREE ((uint8_t *) literal_p->u.char_p);
+      mem_heap_free_block_size_stored ((uint8_t *) literal_p->u.char_p);
```
Comment (akiss77) [inline jerry-core/parser/js/common.c]: Better cast this to `void *`

File: jerry-core/parser/js/js-parser.c [hunk 16]
```diff
@@ -105,7 +105,7 @@ parser_compute_indicies (parser_context_t *context_p, /**< context */
 
         if (!(literal_p->status_flags & LEXER_FLAG_SOURCE_PTR))
         {
-          PARSER_FREE ((uint8_t *) char_p);
+          mem_heap_free_block_size_stored ((uint8_t *) char_p);
```
Comment (akiss77) [inline jerry-core/parser/js/js-parser.c]: ditto

**Comment by akiss77** [comment]
After the casts are fixed, LGTM (FWIW)

File: jerry-core/jerry.c (hunk 5 above)
Comment (franc0is) [inline jerry-core/jerry.c]: It's fine if this is just not acceptable style for this project. I was only pointing it out because I recently had to fix a pretty nasty regression where a field was added to a struct that was initialized field by field as you do here and ended up uninitialized:

```
typedef struct {
  int a;
  int b;
  int c;
} foo_t;
...
foo_t bar;
// when this code was written, foo_t only had two fields, a and b.
bar.a = 1;
bar.b = 2;
...
// bar.c is uninitialized
```

**Comment by zherczeg** [comment]
I hope I fixed everything.

**Comment by LaszloLango** [comment]
LGTM
